pkg/hubble/metrics/dynamic_flow_processor.go

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package metrics
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+
+	"github.com/cilium/cilium/pkg/hubble/metrics/api"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// DynamicFlowProcessor represents instance of hubble exporter with dynamic
+// configuration reload.
+type DynamicFlowProcessor struct {
+	logger  logrus.FieldLogger
+	watcher *metricConfigWatcher
+	// Protects against deregistering metric handlers while ProcessFlow is executing, or concurrent config reloads.
+	mutex    lock.RWMutex
+	Metrics  []api.NamedHandler
+	registry *prometheus.Registry
+}
+
+// OnDecodedEvent distributes events across all managed exporters.
+func (d *DynamicFlowProcessor) OnDecodedFlow(ctx context.Context, flow *flowpb.Flow) (bool, error) {
+	select {
+	case <-ctx.Done():
+		return false, nil
+	default:
+	}
+
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+
+	var errs error
+	if d.Metrics != nil {
+		errs = api.ExecuteAllProcessFlow(ctx, flow, &d.Metrics)
+	}
+
+	if errs != nil {
+		d.logger.WithError(errs).Error("Failed to ProcessFlow in metrics handler")
+	}
+	return false, errs
+}
+
+// NewDynamicFlowProcessor creates instance of dynamic hubble flow exporter.
+func NewDynamicFlowProcessor(reg *prometheus.Registry, logger logrus.FieldLogger, configFilePath string) *DynamicFlowProcessor {
+	dynamicFlowProcessor := &DynamicFlowProcessor{
+		logger:   logger,
+		registry: reg,
+	}
+	watcher := NewMetricConfigWatcher(configFilePath, dynamicFlowProcessor.onConfigReload)
+	dynamicFlowProcessor.watcher = watcher
+	return dynamicFlowProcessor
+}
+
+func (d *DynamicFlowProcessor) onConfigReload(ctx context.Context, hash uint64, config api.Config) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	var newHandlers []api.NamedHandler
+	metricNames := config.GetMetricNames()
+
+	curHandlerMap := make(map[string]*api.NamedHandler)
+	if d.Metrics != nil {
+		for _, m := range d.Metrics {
+			curHandlerMap[m.Name] = &m
+		}
+
+		configuredMetricNames := make(map[string]*api.MetricConfig)
+		for _, cm := range config.Metrics {
+			configuredMetricNames[cm.Name] = cm
+		}
+		// Unregister handlers not present in the new config.
+		// This needs to happen first to properly check for conflicting plugins later during registration.
+		for _, m := range d.Metrics {
+			if _, ok := configuredMetricNames[m.Name]; !ok {
+				h := curHandlerMap[m.Name]
+				err := h.Handler.Deinit(d.registry)
+				if err != nil {
+					d.logger.WithField("name", m.Name).WithError(err).Error("Deinit failed for handler")
+				}
+				delete(curHandlerMap, m.Name)
```

Comment:
Error handling is tricky. If a plugin fails to `Deinit` then it might be in an unclean state and then we might not want to proceed with updating the metrics config. However, at least with the currently written plugins, they only handle unregistering metrics so this is probably not an issue at the moment, but I worry about the repercussions if they ever have some kind of state.

Can you think of any good ways to handle potential error cases?

Reply:
I'd say the handlers are already stateful because they hold the config and metric vectors, so do you mean a situation where they hold state across re-registrations?

IMO a handler that failed to Deinit can be in an invalid state and we don't want to rely on it for reporting, so it should be removed from the map.

At the same time, if a failed Deinit leaves a handler in a state that still generates metrics, then perhaps it shouldn't be removed from the map here - we could decide based on the specific error returned. 

The question of transactionality of config changes for a failed Deinit seems the same as for a failed update below - I'll address it in your other comment.